src/site/content/en/blog/kapwing/index.md

```diff
@@ -0,0 +1,415 @@
+information performantly and to update quickly on a zoom or pan of the
+timeline.
+
+The snippet below demonstrates how we do this:
+
+```js
+const getDownsampledBuffer = (idbAsset: IdbAsset) =>
+  decodeMutex.runExclusive(
+    async (): Promise<Float32Array> => {
+      const arrayBuffer = await idbAsset.file.arrayBuffer();
+      const audioContext = new AudioContext();
+      const audioBuffer = await audioContext.decodeAudioData(arrayBuffer);
+
+      const offline = new OfflineAudioContext(
+        audioBuffer.numberOfChannels,
+        audioBuffer.duration * MIN_BROWSER_SUPPORTED_SAMPLE_RATE,
+        MIN_BROWSER_SUPPORTED_SAMPLE_RATE
+      );
+
+      const downsampleSource = offline.createBufferSource();
+      downsampleSource.buffer = audioBuffer;
+      downsampleSource.start(0);
+      downsampleSource.connect(offline.destination);
+
+      const downsampledBuffer22K = await offline.startRendering();
+
+      const downsampledBuffer22KData = downsampledBuffer22K.getChannelData(0);
+
+      const downsampledBuffer = new Float32Array(
+        Math.floor(
+          downsampledBuffer22KData.length / POST_BROWSER_SAMPLE_INTERVAL
+        )
+      );
+
+      for (
+        let i = 0, j = 0;
+        i < downsampledBuffer22KData.length;
+        i += POST_BROWSER_SAMPLE_INTERVAL, j += 1
+      ) {
+        let sum = 0;
+        for (let k = 0; k < POST_BROWSER_SAMPLE_INTERVAL; k += 1) {
+          sum += Math.abs(downsampledBuffer22KData[i + k]);
+        }
+        const avg = sum / POST_BROWSER_SAMPLE_INTERVAL;
+        downsampledBuffer[j] = avg;
+      }
+
+      return downsampledBuffer;
+    } 
+  );
+```
+
+We pass this helper the asset that is stored in indexedDB. Upon completion we 
+will update the asset in indexedDB as well as our own cache.
+
+We gather data about the audioBuffer with the AudioContext constructor, but 
+because we aren't rendering to the device hardware we use the 
+`OfflineAudioContext` to render to an `ArrayBuffer` where we will store 
+amplitude data.
+
+The API itself returns data at a sample rate much higher than needed for 
+effective visualization. That's why we manually downsample to 200 Hz, which we 
+found to be enough for useful, visually appealing waveforms. 
+
+### WebCodecs
+
+For certain videos the track thumbnails are more useful for timeline 
+navigation than the waveforms. However, generating thumbnails is more resource 
+intensive than generating waveforms.
+
+We cannot cache every potential thumbnail on load, so quick decode on timeline 
+pan/zoom is critical to a performant and responsive application. The 
+bottleneck to achieving smooth frame drawing is decoding frames, which until 
+recently we did using an HTML5 video player. The performance of that approach 
+was not reliable and we often saw degraded app responsiveness during frame 
+rendering.
+
+Recently we have moved over to web codecs, which can be used in web workers. 
```

Comment:
Is it supposed to consistently be "WebCodecs" or "web codecs"?

Reply:
Good catch, WebCodecs, we want consistency and precision. Glad to have your sharp eye on this.

Would also be good to add links to web.dev posts for anything that we have covered previously. Like this https://web.dev/webcodecs/

Reply:
Great -- linked out!